allennlp/semparse/worlds/world.py

```diff
@@ -245,71 +276,86 @@ def get_logical_form(self,
-                else:
-                    break
-            if upcoming_higher_order_functions:
-                if 'lambda' in terminal:
-                    terminal = f"({terminal} {partial_logical_forms.pop()})"
-                    num_args -= 1
-                for upcoming_function in upcoming_higher_order_functions:
-                    terminal = f"({upcoming_function} {terminal})"
-            if num_args == 0:
-                partial_logical_forms.append(terminal)
-            else:
-                args = []
-                if len(partial_logical_forms) < num_args:
-                    logger.error("Not enough arguments for: %s", terminal)
-                    logger.error("Partial logical forms were: %s", partial_logical_forms)
-                    logger.error("Action sequence was: %s", action_sequence)
-                    raise ParsingError("Can't produce logical form from action sequence")
-                for _ in range(num_args):
-                    args.append(partial_logical_forms.pop())
-                partial_logical_forms.append("(%s %s)" % (terminal, " ".join(args)))
-        if len(partial_logical_forms) != 1:
-            logger.error("Incomplete action sequence (or parsing error): %s", action_sequence)
-            logger.error("Partial logical forms were: %s", partial_logical_forms)
-            raise ParsingError("Can't produce logical form from action sequence")
-        return partial_logical_forms[0]
+        # Basic outline: we assume that the bracketing that we get in the RHS of each action is the
+        # correct bracketing for reconstructing the logical form.  This is true when there is no
+        # currying in the action sequence.  Given this assumption, we just need to construct a tree
+        # from the action sequence, then output all of the leaves in the tree, with brackets around
+        # the children of all non-terminal nodes.
+
+        remaining_actions = [action.split(" -> ") for action in action_sequence]
+        tree = _TreeNode(remaining_actions[0][1])
+
+        try:
+            remaining_actions = self._construct_node_from_actions(tree,
+                                                                  remaining_actions[1:],
+                                                                  add_var_function)
+        except ParsingError:
+            logger.error("Error parsing action sequence: %s", action_sequence)
+            raise
+
+        if remaining_actions:
+            logger.error("Error parsing action sequence: %s", action_sequence)
+            logger.error("Remaining actions were: %s", remaining_actions)
+            raise ParsingError("Extra actions in action sequence")
+        return str(tree)
+
+    def _construct_node_from_actions(self,
+                                     current_node: _TreeNode,
+                                     remaining_actions: List[List[str]],
+                                     add_var_function: bool) -> List[List[str]]:
```

Comment:
What does `add_var_function` do? You may want to mention this in the docstring below of add a parameters list.

Reply:
This is a private function, and the parameter is documented in the public `get_logical_form`.  I'm not sure it's worth repeating that documentation here.